pkg/controller/machine.go

```diff
@@ -325,7 +350,7 @@ func (c *controller) updateMachineState(machine *v1alpha1.Machine) (*v1alpha1.Ma
 		clone.Labels = make(map[string]string)
 	}
 
-	if _, ok := clone.Labels["node"]; !ok {
+	if n := clone.Labels["node"]; n == "" {
```

Comment:
>The right way is to depend on the second (ok) field. I have read this somewhere. I think we should revert this back.

`if _, ok := clone.Labels["node"]; !ok` would tell us if key exists or not, but I wanted to explicitly check if key doesn't exist but also not set to ""(empty string). As the code [here](https://github.com/gardener/machine-controller-manager/blob/859ef7504d2ff3efa715bfd6c84317c3267de835/pkg/controller/machine.go#L466-L469) sets the label of `node: ""` on machine-object if MCM restarts immediately after `driver.Create()`, which can slip through if we revert the code back to relying on second(ok) argument.

Reply:
>we should also check for: clone.Labels["node"] != machine.Status.Node

Yes, we should have such a check irrespective of the creation/deletion flow, not sure if it falls under the scope of this PR though.

Reply:
we should still be checking for 2nd `ok` argument i believe. If the crash happens immediately after `driver.Create()` then the `"node"` key would not even exist in `clone.Labels`. And this call itself will fail.
It might be worth testing the scenario by intensionally crashing/panicking MCM just after `driver.Create()` returns.

Reply:
Afaik, if the `key` doesn't exist in the map[string]string, then the expression `k = label[key]` should return "", and not fail/panic.

Reply:
I was mainly concerned about [this](https://github.com/gardener/machine-controller-manager/blob/859ef7504d2ff3efa715bfd6c84317c3267de835/pkg/controller/machine.go#L466-L467) check which is missing now.
If clone.Labels == nil, then the following assignment will panic, right?

Reply:
The labels may already be present if the machine is created by machineset controller, but if a machine alone is created then we might be in trouble. WDYT?

Reply:
I think even if the map is set to nil, it should still not panic. I think it should only return the "" value back if the key is not found.

Reply:
```suggestion
	if nodeLabel, exists := clone.Labels["node"]; !exists &&  nodeLabel== "" {
```
How about we check both scenarios explicitly. Just to be safe? I hope the condition i have written is what you have explained.

Reply:
If `node: ""`  label exists on the machine-object, it will not pass the above-suggested check, right? 

We could probably use `!exists ||  nodeLabel== ""` but it seems to have same impact as the already written check.

Reply:
>  think even if the map is set to nil, it should still not panic. I think it should only return the "" value back if the key is not found.
Yes, but the next assignment `clone.Labels["node"] = machine.Status.Node` panics in this case.

Reply:
> Yes, but the next assignment clone.Labels["node"] = machine.Status.Node panics in this case.

I don't think it should panic considering he is initializing the map a step before. Here - https://github.com/gardener/machine-controller-manager/blob/859ef7504d2ff3efa715bfd6c84317c3267de835/pkg/controller/machine.go#L350. 

> but I wanted to explicitly check if key doesn't exist but (I am assuming but you mean and right?) also not set to ""(empty string)

If that's the case i think it would be 
```suggestion
	if nodeLabel, exists := clone.Labels["node"]; !exists && nodeLabel== "" {
```

Or maybe i understood it wrongly. You can take a call on the condition. But i would suggest you to use the exists/ok value also for the condition check. Refer - https://www.digitalocean.com/community/tutorials/understanding-maps-in-go#checking-existence

Reply:
> I don't think it should panic considering he is initializing the map a step before. Here -

ah. missed that it was already there 2 lines above.

Reply:
Had a call with hardik and is resolved from my side.